### PR TITLE
Updated UI-tests to run inside of docker and switched to circleci 2.0

### PR DESCRIPTION
**Pull request by jrbenny35**
This is a big one....

With the help of @EnTeQuAk we have figured out everything needed to run the UI-Tests _inside_ of the docker container. This will allow better team debugging as everyone can run the same tests.

I have also updated the ```circle.yml``` to build, test, deploy, and release the dockerfile as it currently does but using their new 2.0 build system. I also added the appropriate caching so the builds aren't too slow.

Some of the tests were out of date so I made sure they all work well.

This PR also include the API that @EnTeQuAk wrote up for future use. 

The patch suggested here: #6445 is also concluded.

### CODE REVIEW

**Comment by jrbenny35** [comment]
@davehunt Could you take a look at the tests config? i.e.: tox, pytest test files, etc. Thanks!

File: .travis.yml [hunk 1]
```diff
@@ -6,7 +6,6 @@ python:
 
 env:
   global:
-    - DISPLAY=:99.0
```
Comment (eviljeff) [inline .travis.yml]: what did this do?

File: Makefile-docker [hunk 2]
```diff
@@ -67,12 +67,11 @@ help_submake:
 
 initialize_db:
 	rm -rf ./user-media/* ./tmp/*
-	python manage.py reset_db
+	python manage.py reset_db --noinput
 	python manage.py syncdb --noinput
 	python manage.py loaddata initial.json
 	python manage.py import_prod_versions
 	schematic --fake src/olympia/migrations/
-	python manage.py createsuperuser
```
Comment (eviljeff) [inline Makefile-docker]: if you're removing `createsuperuser` from `initialize_db` it needs to be inserted somewhere else -  `initialize` calls `initialize_db` and that's what developers (us!) call to get everything set up at first.

File: requirements/docs.txt [hunk 3]
```diff
@@ -41,13 +41,13 @@ requests==2.18.4 \
     --hash=sha256:6a1b267aa90cac58ac3a765d067950e7dbbf75b1da07e895d1f594193a40a38b \
     --hash=sha256:9c443e7324ba5b85070c4a818ade28bfabedf16ea10206da1132edaa6dda237e
 # setuptools is required by Sphinx
-setuptools==36.4.0 \
-    --hash=sha256:4d54c0bfee283e78609169213f9c075827d5837086f58b588b417b093c23464b \
-    --hash=sha256:2758b0270fe8ceec42f336ee5b411e60dc8579febc27bb3ba9b794dc7f0239ae
+setuptools==36.5.0 \
+    --hash=sha256:ef824aefbd20dc364891836b75a19386dcf2f4235bf7d80531a8517ab29d0602 \
+    --hash=sha256:ce2007c1cea3359870b80657d634253a0765b0c7dc5a988d77ba803fc86f2c64
```
Comment (eviljeff) [inline requirements/docs.txt]: can you rebase to master - all of the requirements changes should already be there

File: setup.cfg [hunk 4]
```diff
@@ -1,5 +1,4 @@
 [tool:pytest]
-addopts = -vs --reuse-db --showlocals --tb=short
```
Comment (eviljeff) [inline setup.cfg]: why is this being removed - we need this for normal testing, right?

File: src/olympia/browse/templates/browse/personas/category_landing.html [hunk 5]
```diff
@@ -9,7 +9,11 @@
 } %}
 
 {% macro sort_link(key) -%}
-  {{ url('browse.personas', category.slug)|urlparams(sort=key) }}
+{% if category %}
+    {{ url('browse.personas', category.slug)|urlparams(sort=key) }}
+{% else %}
+    {{ url('browse.personas')|urlparams(sort=key) }}
+{% endif %}
```
Comment (eviljeff) [inline src/olympia/browse/templates/browse/personas/category_landing.html]: @diox is your "temporary fix" good enough to be committed?

Comment (diox) [reply src/olympia/browse/templates/browse/personas/category_landing.html]: I'd prefer if this went separately, in another PR, with a matching test.

File: tests/ui/requirements.txt [hunk 6]
```diff
@@ -7,3 +8,9 @@ pytest-selenium==1.11.1
 pytest-variables==1.7.0
 pytest-xdist==1.20.0
 selenium==3.5.0
+hashin==0.11.2
```
Comment (davehunt) [inline tests/ui/requirements.txt]: Do we require hashin? Should we include the hashes for each of these packages?

**Comment by davehunt** [comment]
> Could you take a look at the tests config? i.e.: tox, pytest test files, etc. Thanks!

I think this looks great @jrbenny35! I'd love to try running it myself. I'm sure I could work it out, but could you include a bit of documentation for how to do this? I suspect it would help others too.

File: .travis.yml (hunk 1 above)
Comment (EnTeQuAk) [reply .travis.yml]: This was needed to be able to run xfvc, open a browser window (though headless) and render stuff.

File: Makefile-docker [hunk 7]
```diff
@@ -67,12 +67,11 @@ help_submake:
 
 initialize_db:
 	rm -rf ./user-media/* ./tmp/*
-	python manage.py reset_db
+	python manage.py reset_db --noinput
```
Comment (EnTeQuAk) [inline Makefile-docker]: Although it does make sense I liked that `reset_db` asked before destroying all the things. I'd do that manually somewhere if that change is still needed.

File: circle.yml [hunk 8]
```diff
@@ -1,38 +1,109 @@
...
+      - checkout
+      - run: >
+          printf '{"commit":"%s","version":"%s","source":"https://github.com/%s/%s","build":"%s"}\n'
+          "$CIRCLE_SHA1"
+          "$CIRCLE_TAG"
+          "$CIRCLE_PROJECT_USERNAME"
+          "$CIRCLE_PROJECT_REPONAME"
+          "$CIRCLE_BUILD_URL"
+          > version.json
+      - run: docker build -t app:build -f Dockerfile.deploy .
```
Comment (EnTeQuAk) [inline circle.yml]: note to myself: This looks great, should be tested thoughtfully though to make sure we don't break our deploys.

File: circle.yml [hunk 9]
```diff
@@ -1,38 +1,109 @@
...
+    machine: true
+    steps:
+      - checkout
+      - restore_cache:
+          keys:
+            - uitest-cache-{{ checksum "requirements/docs.txt" }}
+            - uitest-cache-{{ checksum "requirements/prod.txt" }}
+            - uitest-cache-{{ checksum "requirements/dev.txt" }}
+            - uitest-cache-{{ checksum "tests/ui/requirements.txt" }}
+            - uitest-cache-
```
Comment (EnTeQuAk) [inline circle.yml]: Guess that's one key too much?

File: Makefile-docker (hunk 2 above)
Comment (jrbenny35) [reply Makefile-docker]: Good point. Should I just put all the initialize stuff for the ui tests in its own make command? I just didn't want to have 2 commands doing the same thing essentially.

File: circle.yml [hunk 10]
```diff
@@ -1,38 +1,109 @@
...
+          paths:
+            - .tox
+      - save_cache:
+          key: uitest-cache-{{ checksum "tests/ui/requirements.txt" }}
+          paths:
+            - .tox
+      - save_cache:
+          key: uitest-cache-
+          paths:
+            - .tox
```
Comment (EnTeQuAk) [inline circle.yml]: Guess that's one key too much again heh

File: requirements/docs.txt (hunk 3 above)
Comment (jrbenny35) [reply requirements/docs.txt]: True. I need to use a certain version of certifi though, as the requirements installed for the selenium tests need a newer version.

File: setup.cfg (hunk 4 above)
Comment (jrbenny35) [reply setup.cfg]: Maybe this was by error, I'll recheck.

File: src/olympia/browse/templates/browse/personas/category_landing.html (hunk 5 above)
Comment (jrbenny35) [reply src/olympia/browse/templates/browse/personas/category_landing.html]: I can write a test and submit another PR.

**Comment by jrbenny35** [comment]
> I think this looks great @jrbenny35! I'd love to try running it myself. I'm sure I could work it out, but could you include a bit of documentation for how to do this? I suspect it would help others too.

I can and I will. For now make sure to pull and update all containers and start them using ```docker-compose up -d``` then run ```docker-compose exec web bash /code/scripts/ui-test.sh```.

File: docker-compose.yml [hunk 11]
```diff
@@ -1,69 +1,74 @@
...
+version: "2"
 
-web:
-  <<: *worker
-  command: supervisord -n -c /code/docker/supervisor.conf
+services:
+  worker: &worker
+    image: addons/addons-server
+    command: supervisord -n -c /code/docker/supervisor-celery.conf
+    entrypoint: ./scripts/start-docker.sh
```
Comment (EnTeQuAk) [inline docker-compose.yml]: Does that work now? 😨

File: requirements/docs.txt [hunk 12]
```diff
@@ -65,11 +65,10 @@ sphinxcontrib-websupport==1.0.1 \
...
     --hash=sha256:9f3bcd3c401c3e862ec0ebe6d2c069ebc012ce142cce209c098ccb5b09136e89 \
     --hash=sha256:919f26a68b2c17a7634da993d91339e288964f93c274f1343e3bbbe2096e1128
-idna==2.6 \
-    --hash=sha256:8c7309c718f94b3a625cb648ace320157ad16ff131ae0af362c9f21b80ef6ec4 \
-    --hash=sha256:2c6a5de3089009e3da7c5dde64a141dbc8551d5b7f6cf4ed7c2568d0cc520a8f
-certifi==0.0.8 \
-     --hash=sha256:46ecf5f7526a08cc1f8bc8232adf0cffce046f46ceff95539daec42ebc4849ef # pyup: ==0.0.8
+idna==2.5 \
+    --hash=sha256:cc19709fd6d0cbfed39ea875d29ba6d4e22c0cebc510a76d6302a28385e8bb70
+certifi==2017.4.17 \
```
Comment (EnTeQuAk) [inline requirements/docs.txt]: We **must not** update certifi, there's a bug with it unfortunately and it breaks certificate related stuff on production. So this should be backed out.

File: requirements/prod.txt [hunk 13]
```diff
@@ -102,8 +102,8 @@ bleach==1.5.0 \
 celery==3.1.23 \
     --hash=sha256:eaf5dee3becbc35c7754a2d4482d53bdf72ea3f85dd258525259983262081474 \
     --hash=sha256:1a359c815837f9dbf193a7dbc6addafa34612c077ff70c66e3b16e14eebd2418 # pyup: ==3.1.23
-certifi==0.0.8 \
-    --hash=sha256:46ecf5f7526a08cc1f8bc8232adf0cffce046f46ceff95539daec42ebc4849ef # pyup: ==0.0.8
+certifi==2017.4.17 \
+    --hash=sha256:f4318671072f030a33c7ca6acaef720ddd50ff124d1388e50c1bda4cbd6d7010
```
Comment (EnTeQuAk) [inline requirements/prod.txt]: Same here, 

We must not update certifi, there's a bug with it unfortunately and it breaks certificate related stuff on production. So this should be backed out.

File: scripts/ui-test.sh [hunk 14]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+echo 127.0.0.1 olympia.dev | tee -a /etc/hosts
+yum -y install curl
+curl https://raw.githubusercontent.com/creationix/nvm/v0.30.2/install.sh > install-nvm.sh
+sh install-nvm.sh
+source ~/.bash_profile
+nvm install node
+unset NPM_CONFIG_PREFIX
+pip install tox
+tox -e ui-tests
```
Comment (EnTeQuAk) [inline scripts/ui-test.sh]: So neat this finally works with nvm, great stuff!

File: setup.cfg (hunk 4 above)
Comment (EnTeQuAk) [reply setup.cfg]: Agreed, that change should be backed out, we need that for our regular tests. The test results on travis are useless without those options.

File: src/olympia/landfill/management/commands/generate_ui_test_addons.py [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+    """
+
+    def handle(self, *args, **kwargs):
+        translation.activate('en-US')
+        serializer = GenerateAddonsSerializer()
+        serializer.create_addons()
+        serializer.create_addon()
+        serializer.create_theme()
+        serializer.create_collections()
+        serializer.create_themes()
```
Comment (EnTeQuAk) [inline src/olympia/landfill/management/commands/generate_ui_test_addons.py]: This looks like things are a bit duplicate here and there. You're already calling `create_addons` and then again call `create_addon` so either you should give those methods proper names if they're creating different kinds of data (e.g the first might be just n dummy add-ons while the second one adds a very full featured add-on).

Worth some cleanup.

File: src/olympia/landfill/serializers.py [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+                tags=['some_tag', 'another_tag', 'ui-testing',
+                      'selenium', 'python'],
+                total_reviews=777,
+                weekly_downloads=22233879,
+                developer_comments='This is a testing addon.',
+            )
+            Preview.objects.create(addon=addon, position=1)
+            Review.objects.create(addon=addon, rating=5, user=user_factory())
+            Review.objects.create(addon=addon, rating=3, user=user_factory())
+            addon.reload()
```
Comment (EnTeQuAk) [inline src/olympia/landfill/serializers.py]: I don't think `addon.reload` should be called, the object is not being used afterwards where we'd need that.

File: src/olympia/landfill/serializers.py [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+
+    def create_addon(self):
+        """Creates a custom addon named 'Ui-Addon'.
+
+        This addon will be a featured addon and will have a featured collecton
+        attatched to it. It will belong to the user uitest.
+
+        It has 1 preview, 5 reviews, and 2 authors. The second author is named
+        'ui-tester2'. It has a version number as well as a beta version.
+
```
Comment (EnTeQuAk) [inline src/olympia/landfill/serializers.py]: Given that docstring the method should be called `create_featured_addon_with_version` or so to at least reflect that.

File: src/olympia/landfill/serializers.py [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+            'Created addon {0} for testing successfully'
+            .format(addon.name))
+
+    def create_theme(self):
+        """Creates a custom theme named 'Ui-Test Theme'.
+
+        This theme will be a featured theme and will belong to the uitest user.
+
+        It has one author.
+
```
Comment (EnTeQuAk) [inline src/olympia/landfill/serializers.py]: The method should be renamed a bit to reflect it's more full-featured implementation, something like `create_featured_theme_with_data` or so?

File: src/olympia/landfill/views.py [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+            call_command('flush', verbosity=0, interactive=False,
+                         database=db_name, reset_sequences=False,
+                         allow_cascade=False,
+                         inhibit_post_migrate=False)
+
+        # print('Restore...')
+        # Restore database to previously known state
+        connection.creation.deserialize_db_from_string(data)
+
+        return Response({}, status=200)
```
Comment (EnTeQuAk) [inline src/olympia/landfill/views.py]: If we don't use the views for now we should just delete them until we figure out a way to fix them and make them work.

File: tests/ui/conftest.py [hunk 20]
```diff
@@ -1,50 +1,17 @@
...
-    ADDON_PERSONA,
-    STATUS_PUBLIC
-)
-from olympia.landfill.collection import generate_collection
-from olympia.landfill.generators import generate_themes
-from olympia.reviews.models import Review
-from olympia.users.models import UserProfile
-from pytest_django import live_server_helper
 
 # TODO: Change imports to allow running tests against deployed instances.
```
Comment (EnTeQuAk) [inline tests/ui/conftest.py]: I think you just fixed that comment :)

File: tests/ui/requirements.txt (hunk 6 above)
Comment (EnTeQuAk) [reply tests/ui/requirements.txt]: Imho, yes, we should start using hashes for the ui-test requirements file. Imho we could also move that file over to the `requirements` folder in the root folder to better reflect that it's there and make it easier to scan what dependencies we use for all our code.

File: tests/ui/requirements.txt [hunk 21]
```diff
@@ -7,3 +8,9 @@ pytest-selenium==1.11.1
...
+zope.component==4.3.0
+zope.event==4.2.0
+zope.interface==4.3.3
```
Comment (EnTeQuAk) [inline tests/ui/requirements.txt]: What are they used for? Or rather, for what dependency did you add them?

File: circle.yml (hunk 9 above)
Comment (jrbenny35) [reply circle.yml]: Heh, so this means that the cache will not be restored if any of those files change. 

See: https://circleci.com/docs/2.0/caching/

File: circle.yml (hunk 10 above)
Comment (jrbenny35) [reply circle.yml]: https://circleci.com/docs/2.0/caching/

File: docker-compose.yml (hunk 11 above)
Comment (jrbenny35) [reply docker-compose.yml]: Yes! Remember it was the fact that we couldn't mount within a docker image.

File: requirements/docs.txt (hunk 12 above)
Comment (jrbenny35) [reply requirements/docs.txt]: I need certifi to be updated as one of the selenium test requirements needs to be on at least this version. I'll do some research.

File: Makefile-docker (hunk 2 above)
Comment (EnTeQuAk) [reply Makefile-docker]: I think I'd prefer that. It's separate enough that creating a new command kinda makes sense to me

File: tests/ui/requirements.txt (hunk 21 above)
Comment (jrbenny35) [reply tests/ui/requirements.txt]: I think I was trying to pin a bunch of them to figure out some early problems.

File: circle.yml (hunk 9 above)
Comment (EnTeQuAk) [reply circle.yml]: But that last key does not reference any file 😕

File: circle.yml (hunk 10 above)
Comment (EnTeQuAk) [reply circle.yml]: Same here, the key does not reference any file 😕

File: docker-compose.yml (hunk 11 above)
Comment (EnTeQuAk) [reply docker-compose.yml]: Ha, I do! Completely forgot about that 😆

File: requirements/docs.txt (hunk 12 above)
Comment (EnTeQuAk) [reply requirements/docs.txt]: hmm, that's not good. Then we should **only** use a different version for the ui-tests and don't touch any of the regular requirements files. Eventually we should upgrade certifi in prod anyway but that's a hard process unfortunately.

File: circle.yml (hunk 9 above)
Comment (jrbenny35) [reply circle.yml]: If the cache gets changed it updates the key, so when it restores the cache, even if one of the files are not changed the cache stored at key would have been changed. Confusing kinda but the docs explain it better.

File: requirements/prod.txt (hunk 13 above)
Comment (jrbenny35) [reply requirements/prod.txt]: What do you think of adding a pip install just for certifi in the Makefile? I can't get around the ```Multiple .dist-info directories``` error.

Comment (EnTeQuAk) [reply requirements/prod.txt]: As a workaround, yeah, let's do that for now but mark it with a big `TODO` and open an issue to investigate alternatives.

**Comment by jrbenny35** [comment]
r? @EnTeQuAk 

I had to add the install certifi command to tox. Very temp. solution.

**Comment by EnTeQuAk** [comment]
Thanks for the update! I don't really like the certifi situation but I don't have a better solution right now. Maybe @diox has some ideas. An updated certifi doesn't bother us as much in local development as it does on dev, stage or prod so I'm fine with ui-tests relying on a more recent version.

**Comment by EnTeQuAk** [comment]
We finally upgraded certifi now and are testing it out on -dev.

@jrbenny35 mind removing the certifi hack again? Sorry about all the back and forth :(

Otherwise this looks great so once the hack is removed I think we can go ahead and merge it and improve continuously :)

**Comment by EnTeQuAk** [comment]
@jrbenny35 any update on removing the certifi hack again so that we can get this merged? Imho that's the last piece blocking this PR.

**Comment by EnTeQuAk** [comment]
I merged this with master and reverted one minor version downgrade in requirements/prod.txt.

Once tests pass I'm going to merge this.

**Comment by EnTeQuAk** [comment]
@jrbenny35 I just noticed that we don't have any proper, fine grained github-issues for the ui-tests. Let's create some and fix them one by one from now on instead of working on monster pull requests.

Do you have a few issues in mind that you can create?

**Comment by jrbenny35** [comment]
> @jrbenny35 I just noticed that we don't have any proper, fine grained github-issues for the ui-tests. Let's create some and fix them one by one from now on instead of working on monster pull requests.

Do you have a few issues in mind that you can create?

Yeah I do, I'll start adding them.

**Pull request merged**